### PR TITLE
[CM-24]Sanitize file input for parser

### PR DESCRIPTION
**Pull request by naveen-nathan**
### Jira Ticket

[Jira Ticket](https://dashbd.atlassian.net/browse/CM-24)

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would change existing functionality)

### Checklist

- [X] My branch name matches the format: `<ticket-id>/<brief-description-of-change>`
- [X] My PR name matches the format: `[<ticket-id>] <brief-description-of-change>`
- [ ] I have added doc-comments to all new functions ([JSDoc](https://jsdoc.app/) for JS and [Docstrings](https://peps.python.org/pep-0257/) for Python)
- [X] I have reviewed all of my code
- [X] My code only contains major changes related to my ticket

### CODE REVIEW

File: progressReport/parser.py [hunk 1]
```diff
@@ -138,9 +141,96 @@ def nodes_to_json(node):
...
+                },
+                "required": ["id", "name", "parent", "children", "data"]
+            }
+        }
+    }
+    try:
+        validate(instance=data, schema=schema)
+    except ValidationError as e:
+        # Not sure how we should handle the error, so just raising it for now.
+        raise e
```
Comment (naveen-nathan) [inline progressReport/parser.py]: Not sure how we should handle the error, so just raising it for now.

**Comment by naveen-nathan** [comment]
This was implemented assuming that the file users upload would be in json format as opposed to a text file in [this](https://github.com/AFA-Tooling/GradeView/blob/main/progressReport/meta/Berkeley_CS10.txt) format (explanation in [this Slack message](https://dancsed.slack.com/archives/C07M38E2ZQR/p1731287928602489)). If this assumption isn’t true, happy to change the PR accordingly!

File: progressReport/parser.py (hunk 1 above)
Comment (Connor-Bernard) [reply progressReport/parser.py]: I would not try catch it for now, just let it be thrown so that we can handle it as necessary during later invocations

**Review by Connor-Bernard — COMMENTED** [review]
wait we need to sanitize the schema file not the output json

**Comment by Connor-Bernard** [comment]
> wait we need to sanitize the schema file not the output json

Right because the issue is if the parser may receive some unclean input which could cause it to be vulnerable.

**Comment by naveen-nathan** [comment]
> > wait we need to sanitize the schema file not the output json
> 
> Right because the issue is if the parser may receive some unclean input which could cause it to be vulnerable.

I was initially hoping that we could have the text-to-json code on the client side, so that we wouldn’t need to worry about an attack targeting the text-to-json code (context [here](https://dancsed.slack.com/archives/C07M38E2ZQR/p1731287928602489)). However, I now realize that it won’t be too much effort to sanitize the initial text file, and will make changes accordingly

File: progressReport/parser.py (hunk 1 above)
Comment (naveen-nathan) [reply progressReport/parser.py]: Noted!

**Comment by naveen-nathan** [comment]
> > wait we need to sanitize the schema file not the output json
> 
> Right because the issue is if the parser may receive some unclean input which could cause it to be vulnerable.

I have now added injection and DDOS prevention for the schema file

**Review by Connor-Bernard — APPROVED** [review]
LGTM

**Pull request merged**